In [1]:
import pandas as pd
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt




In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv")

In [3]:
df.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

>## Question 1

how to merge two lists

In [8]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]
A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

> ## Question 2

How to create an identity matrix

In [9]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

> ## Question 3 

which fuel type has the least average fuel cost per unit burned

In [27]:
df.groupby('fuel_type_code_pudl').fuel_cost_per_unit_burned.mean().sort_values()

fuel_type_code_pudl
gas          13.659397
other        18.253856
waste        19.518122
coal         67.421830
oil         168.877086
nuclear    4955.157002
Name: fuel_cost_per_unit_burned, dtype: float64

>## Question 4 

Calculation he standard deviation and 75th

In [15]:
#standard deviation pf the energy per unit

std = df.fuel_mmbtu_per_unit.std()
std =round(std,2)

#75th percentile of the energy per unit
percentile = df.fuel_mmbtu_per_unit.quantile(0.75)
percentile = round(percentile,2)

print(std,percentile)

10.6 17.01


The standard deviaion and 75th percentile are 10.6 and 17.01 respectively to 
2 d.p

> ## Question 5

calculate skewness and kurtosis of the data to 2 decimal places

In [16]:
#the skewness of the fuel quantity burned

skew = round(df.fuel_qty_burned.skew(),2)

#the kurtosis of the fuel quantity burned
kurt = round(df.fuel_qty_burned.kurtosis(),2)

print(skew, kurt)

15.85 651.37


The skewness and Kurtosis are 15.85 and 651.37 respectively

> ## Question 6

Missing value in the dataset

In [19]:
#missing value

df.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [20]:
#percentage of the missing value in the other rows

sum(df.isna().sum()/len(df))*100

0.609694136774718

Percentage = 0.61% to 3 d.p.

> ## Question 7

Which type of imputation would you use for the missing

In [21]:
#to know the type of value imputation to use we need to know the type of data
df.fuel_unit
#the datatype of its content is 'object' ,implying strings, so categorical and 
# mode imputations are better suited


0        ton
1        ton
2        ton
3        ton
4        mcf
        ... 
29518    mcf
29519    mcf
29520    mcf
29521    ton
29522    mcf
Name: fuel_unit, Length: 29523, dtype: object

>## Question 8

To find the correlation between Fuel cost per unit burned and other features

In [28]:
corr = df.corr()['fuel_cost_per_unit_burned'].sort_values()
corr

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

It can be seen that fuel_qty_burned and fuel_mmbtu_per_unit are the seconf and third lowest

>## Question 9

Percentage change in the fuel cost per unit burned from 1994 to 1998

In [24]:
a = df.fuel_cost_per_unit_burned[(df.fuel_type_code_pudl =='coal')& (df.report_year == 1994)].sum()

b = df.fuel_cost_per_unit_burned[(df.fuel_type_code_pudl =='coal')& (df.report_year == 1998)].sum()

c =((b-a)/a) * 100
c

-20.56765451826053

Percentage change is -20.567 %  (~ -21%)

> ## Question 10

To calculate the year with the highest cost per unit delivered

In [26]:
df.groupby('report_year').fuel_cost_per_unit_delivered.mean().sort_values()

report_year
1999       25.551627
1995       32.735269
2006       38.657484
2005       41.438184
2007       43.325023
2017       46.196861
2002       47.594361
2003       55.663493
2008       58.588197
2011       59.774667
2001       60.050396
2012       60.994502
1994       63.636060
2010       91.862105
2016      103.901761
2004      139.524275
2013      172.307591
2014      192.737183
1998      287.154420
2015      326.535511
2018      499.269966
2009      652.694163
2000      985.362877
1996     9196.705948
1997    11140.197239
Name: fuel_cost_per_unit_delivered, dtype: float64

the year with the highest is 1997